libraries/AP_OpenDroneID/AP_OpenDroneID_DroneCAN.cpp

```diff
@@ -0,0 +1,249 @@
+{
+    const uint8_t driver_index = uavcan->get_driver_index();
+    const uint8_t driver_mask = 1U<<driver_index;
+
+    if (driver_index+1 != _can_driver) {
+        // not enabled for this CAN driver
+        return;
+    }
+
+    dronecan_init(uavcan);
+
+    if (dronecan_init_failed & driver_mask) {
+        return;
+    }
+
+    if (need_send_location & driver_mask) {
+        WITH_SEMAPHORE(_sem);
+        dronecan_send_location(uavcan);
+        need_send_location &= ~driver_mask;
+    }
+    if (need_send_basic_id & driver_mask) {
+        WITH_SEMAPHORE(_sem);
+        dronecan_send_basic_id(uavcan);
+        need_send_basic_id &= ~driver_mask;
+    }
+    if (need_send_system & driver_mask) {
+        WITH_SEMAPHORE(_sem);
+        dronecan_send_system(uavcan);
+        need_send_system &= ~driver_mask;
+    }
+    if (need_send_self_id & driver_mask) {
+        WITH_SEMAPHORE(_sem);
+        dronecan_send_self_id(uavcan);
+        need_send_self_id &= ~driver_mask;
+    }
+    if (need_send_operator_id & driver_mask) {
+        WITH_SEMAPHORE(_sem);
+        dronecan_send_operator_id(uavcan);
+        need_send_operator_id &= ~driver_mask;
+    }
+}
+
+#define ODID_COPY(name) msg.name = pkt.name
+#define ODID_COPY_STR(name) do { for (uint8_t i = 0; i<sizeof(pkt.name) && pkt.name[i]; i++) msg.name.push_back(pkt.name[i]); } while(0)
+
+
+void AP_OpenDroneID::dronecan_send_location(AP_UAVCAN *uavcan)
+{
+    dronecan::remoteid::Location msg {};
+    const auto &pkt = pkt_location;
+    ODID_COPY_STR(id_or_mac);
+    ODID_COPY(status);
+    ODID_COPY(direction);
+    ODID_COPY(speed_horizontal);
+    ODID_COPY(speed_vertical);
+    ODID_COPY(latitude);
+    ODID_COPY(longitude);
+    ODID_COPY(altitude_barometric);
+    ODID_COPY(altitude_geodetic);
+    ODID_COPY(height_reference);
+    ODID_COPY(height);
+    ODID_COPY(horizontal_accuracy);
+    ODID_COPY(vertical_accuracy);
+    ODID_COPY(barometer_accuracy);
+    ODID_COPY(speed_accuracy);
+    ODID_COPY(timestamp);
+    ODID_COPY(timestamp_accuracy);
+    dc_location[uavcan->get_driver_index()]->broadcast(msg);
+}
+
+void AP_OpenDroneID::dronecan_send_basic_id(AP_UAVCAN *uavcan)
+{
+    dronecan::remoteid::BasicID msg {};
+    const auto &pkt = pkt_basic_id;
+    ODID_COPY_STR(id_or_mac);
+    ODID_COPY(id_type);
+    ODID_COPY(ua_type);
+    ODID_COPY_STR(uas_id);
+    dc_basic_id[uavcan->get_driver_index()]->broadcast(msg);
+}
+
+void AP_OpenDroneID::dronecan_send_system(AP_UAVCAN *uavcan)
+{
+    dronecan::remoteid::System msg {};
+    const auto &pkt = pkt_system;
+    ODID_COPY_STR(id_or_mac);
+    ODID_COPY(operator_location_type);
+    ODID_COPY(classification_type);
+    ODID_COPY(operator_latitude);
+    ODID_COPY(operator_longitude);
+    ODID_COPY(area_count);
+    ODID_COPY(area_radius);
+    ODID_COPY(area_ceiling);
+    ODID_COPY(area_floor);
+    ODID_COPY(category_eu);
+    ODID_COPY(class_eu);
+    ODID_COPY(operator_altitude_geo);
+    ODID_COPY(timestamp);
```

Comment:
I think you now assume that the system message come from the GCS. That is fine. The below is mainly FYI.

The GCS needs to fill this timestamp field and the time needs to be valid, since it is used by receivers to validate also the timing of the location data.
The FAA did not approve that the shortened timestamp field in the Location message is enough to meet their rule requirement. This full 32-bit timestamp field was therefore added to the System message and the Means of Compliance document updated to mandate that the System message is mandatory, must be sent once per second and this timestamp field must be filled.
In praxis, this means the GCS needs to send the System MAVLink message to the flight controller once per second.

Reply:
the transmitter module can correct the timestamp based on the last msg from the flight controller plus its own internal clock. That should be ok I think?

Reply:
> the transmitter module can correct the timestamp based on the last msg from the flight controller plus its own internal clock. That should be ok I think?

In theory possible, but then you pose a requirement on the transmitter implementation. Where should this be documented and how to ensure that all transmitter implementations will do this, since it is more like an exception that the transmitter suddenly needs to start modifying the data that is being transmitted.

So far, the assumption has been that the GCS+flight controller generate all the valid message data and the transmitter is just a "dumb" device that blurbs that out into the aether (?).

Probably you are in a better situation to decide what would make sense from a system architecture/eco-system perspective:
1) The GCS updates the System timestamp regularly and publishes new MAVLink messages regularly (for the dynamic Operator Location case, it would anyway need to do this)
2) The flight controller increases the frequency of the System messages that it publishes and ensures the timestamp is up-to-date
3) The transmitter overwrites the System timestamp field to make sure it is up-to-date before transmitting it on the air